File: backend/app/services/segment_service.py
"""
Generic XBRL-style segment data parser, validator, and normalizer.

Handles multi-dimensional financial segment data where:
  - item.name       = metric  (e.g. "us-gaap:RevenueFromContractWithCustomer…")
  - segment.axis    = dimension  (e.g. "srt:ProductOrServiceAxis")
  - segment.key     = stable XBRL member key  (e.g. "srt:Services")
  - segment.label   = human-readable display name  (e.g. "Services")

Nothing in this module is company- or label-specific.
"""
from __future__ import annotations

import logging
from dataclasses import dataclass

log = logging.getLogger(__name__)

# Tolerance for aggregate-sum detection (1 %)
_AGG_TOLERANCE = 0.01


# ── Output type ───────────────────────────────────────────────────────────────

@dataclass
class ParsedSegment:
    key:    str
    label:  str
    value:  float
    axis:   str
    metric: str
    period: str


# ── Core parser ───────────────────────────────────────────────────────────────

def parse_segments(
    periods:     list[dict],
    metric_name: str,
    axis:        str,
    period:      str | None = None,   # report_period string; default: latest
    dedupe:      bool       = True,
    leaf_only:   bool       = True,
) -> list[ParsedSegment]:
    """
    Extract clean, chart-ready segment data from raw segmented revenue periods.

    Parameters
    ----------
    periods     : raw list of SegmentedRevenuePeriod dicts from the FD API
    metric_name : XBRL item name to filter by (item["name"])
    axis        : XBRL axis to filter by (segment["axis"])
    period      : specific report_period to target; defaults to the latest
    dedupe      : skip duplicate segment.key entries within the same period
    leaf_only   : exclude aggregate rows (items with 0 or >1 segments,
                  and items whose value ≈ sum of all sibling values)

    Returns
    -------
    List of ParsedSegment — one entry per leaf node in the chosen period.
    """
    if not periods:
        return []

    # Sort DESC by report_period, pick target
    sorted_periods = sorted(
        periods,
        key=lambda p: p.get("report_period", ""),
        reverse=True,
    )
    if period:
        target = next(
            (p for p in sorted_periods if p.get("report_period") == period), None
        )
        if target is None:
            log.warning("parse_segments: period %r not found", period)
            return []
    else:
        target = sorted_periods[0]

    report_period = target.get("report_period", "")
    seen_keys: set[str] = set()
    results: list[ParsedSegment] = []

    for item in target.get("items", []):
        # ── 1. Filter by metric ────────────────────────────────────────────────
        if item.get("name") != metric_name:
            continue

        segs = item.get("segments", [])

        # ── 2. Filter by axis + enforce leaf constraint ────────────────────────
        if leaf_only:
            # Items with no segments are dimension-less totals → aggregate
            if len(segs) == 0:
                continue
            # Items spanning multiple segments (cross-dimensional) → aggregate
            if len(segs) != 1:
                continue
            if segs[0].get("axis") != axis:
                continue
        else:
            # Without leaf_only: still require exactly one segment on this axis
            axis_segs = [s for s in segs if s.get("axis") == axis]
            if len(axis_segs) != 1:
                continue

        seg     = segs[0]
        seg_key = seg.get("key", "")
        if not seg_key:
            log.debug(
                "parse_segments: missing key for metric=%s axis=%s label=%s",
                metric_name, axis, seg.get("label", "?"),
            )
            continue

        # ── 3. Deduplication by key (NEVER by label) ──────────────────────────
        if dedupe and seg_key in seen_keys:
            log.debug("parse_segments: dedupe skip key=%s", seg_key)
            continue
        seen_keys.add(seg_key)

        results.append(ParsedSegment(
            key    = seg_key,
            label  = seg.get("label") or seg_key,
            value  = float(item.get("amount", 0)),
            axis   = axis,
            metric = metric_name,
            period = report_period,
        ))

    # ── 4. Remove aggregate totals ─────────────────────────────────────────────
    if leaf_only:
        results = _remove_sum_aggregates(results)

    return results


def _remove_sum_aggregates(segs: list[ParsedSegment]) -> list[ParsedSegment]:
    """
    Remove any segment whose value ≈ sum of all other segments.
    These are total rows that slipped through the leaf_only filter because
    they still carry a single segment tag (e.g. the "all products" member).
    """
    if len(segs) < 2:
        return segs

    total = sum(s.value for s in segs)
    cleaned: list[ParsedSegment] = []
    for seg in segs:
        others_sum = total - seg.value
        if others_sum > 0 and abs(seg.value - others_sum) / others_sum < _AGG_TOLERANCE:
            log.debug(
                "parse_segments: removing aggregate key=%s value=%.0f ≈ sum_of_others=%.0f",
                seg.key, seg.value, others_sum,
            )
            continue
        cleaned.append(seg)
    return cleaned


# ── Discovery helpers ─────────────────────────────────────────────────────────

def discover_axes(periods: list[dict]) -> list[str]:
    """Return all unique axis identifiers present in the latest period."""
    if not periods:
        return []
    latest = _latest_period(periods)
    axes: set[str] = set()
    for item in latest.get("items", []):
        for seg in item.get("segments", []):
            ax = seg.get("axis", "")
            if ax:
                axes.add(ax)
    return sorted(axes)


def discover_metrics(periods: list[dict]) -> list[str]:
    """Return all unique metric names present in the latest period."""
    if not periods:
        return []
    latest = _latest_period(periods)
    metrics: set[str] = set()
    for item in latest.get("items", []):
        name = item.get("name", "")
        if name:
            metrics.add(name)
    return sorted(metrics)


def dominant_metric(periods: list[dict], axis: str) -> str | None:
    """
    Return the metric name that has the most leaf items for *axis*.
    Used to auto-select the primary metric when the caller doesn't specify one.
    """
    if not periods:
        return None
    latest = _latest_period(periods)
    counts: dict[str, int] = {}
    for item in latest.get("items", []):
        segs = item.get("segments", [])
        if len(segs) == 1 and segs[0].get("axis") == axis:
            name = item.get("name", "")
            if name:
                counts[name] = counts.get(name, 0) + 1
    if not counts:
        return None
    return max(counts, key=lambda k: counts[k])


def _latest_period(periods: list[dict]) -> dict:
    return max(periods, key=lambda p: p.get("report_period", ""))


# ── Validation ────────────────────────────────────────────────────────────────

def validate_segments(periods: list[dict], ticker: str = "") -> None:
    """
    Log warnings for data anomalies.  Called before caching so issues
    surface early and don't silently corrupt the frontend.

    Checks:
      • duplicate segment.key within the same (metric, axis) context
      • items spanning multiple axes (informational — not an error)
      • missing segment keys
    """
    if not periods:
        return

    for period_data in periods:
        rp    = period_data.get("report_period", "?")
        items = period_data.get("items", [])

        # (metric, axis, key) → occurrence count
        seen: dict[tuple[str, str, str], int] = {}
        missing_key_count = 0

        for item in items:
            metric = item.get("name", "")
            for seg in item.get("segments", []):
                key = seg.get("key", "")
                if not key:
                    missing_key_count += 1
                    continue
                ctx = (metric, seg.get("axis", ""), key)
                seen[ctx] = seen.get(ctx, 0) + 1

        for (metric, ax, key), count in seen.items():
            if count > 1:
                log.warning(
                    "segments [%s %s]: duplicate (metric=%s, axis=%s, key=%s) ×%d",
                    ticker, rp, metric, ax, key, count,
                )

        if missing_key_count:
            log.warning(
                "segments [%s %s]: %d segment(s) have no key",
                ticker, rp, missing_key_count,
            )

        multi_axis = [
            item for item in items
            if len({s.get("axis", "") for s in item.get("segments", [])}) > 1
        ]
        if multi_axis:
            log.debug(
                "segments [%s %s]: %d cross-dimensional items",
                ticker, rp, len(multi_axis),
            )

File: frontend/src/lib/segment_service.ts
/**
 * Generic XBRL-style segment data parser.
 *
 * Handles multi-dimensional financial segment data where:
 *   item.name       = metric  (e.g. "us-gaap:RevenueFromContractWithCustomer…")
 *   segment.axis    = dimension  (e.g. "srt:ProductOrServiceAxis")
 *   segment.key     = stable XBRL member key  (e.g. "srt:Services")
 *   segment.label   = human-readable display name  (e.g. "Services")
 *
 * Nothing in this module is company- or label-specific.
 * All uniqueness guarantees are based on segment.key, NEVER segment.label.
 */

import type { SegmentedRevenuePeriod } from "@/lib/api";

// Tolerance for aggregate-sum detection (1%)
const AGG_TOLERANCE = 0.01;

/** Stable synthetic key for the "Other" collapsed bucket */
export const OTHER_KEY = "__other__";

// ── Output type ───────────────────────────────────────────────────────────────

export interface ParsedSegment {
  key:    string;   // segment.key  — stable XBRL identifier, use as React key
  label:  string;   // segment.label — display only, never use as key
  value:  number;   // item.amount
  axis:   string;   // segment.axis
  metric: string;   // item.name
  period: string;   // report_period of the source row
}

export interface ParseOptions {
  metric_name: string;          // required: filter items by item.name
  axis:        string;          // required: filter segments by segment.axis
  period?:     string;          // target report_period; default: latest
  dedupe?:     boolean;         // deduplicate by segment.key; default true
  leaf_only?:  boolean;         // exclude aggregates; default true
}

// ── Core parser ───────────────────────────────────────────────────────────────

/**
 * Extract clean, chart-ready segment data from raw segmented revenue periods.
 *
 * - Filters by metric_name (item.name) and axis (segment.axis)
 * - Excludes aggregate rows (items with 0 or >1 segments)
 * - Deduplicates by segment.key (NEVER by label)
 * - Removes "grand total" rows (value ≈ sum of all siblings)
 * - Returns the latest period unless `options.period` is specified
 */
export function parseSegments(
  periods: SegmentedRevenuePeriod[],
  options: ParseOptions,
): ParsedSegment[] {
  const { metric_name, axis, period, dedupe = true, leaf_only = true } = options;

  if (!periods.length) return [];

  // Sort DESC by report_period, pick target
  const sorted = [...periods].sort((a, b) =>
    b.report_period.localeCompare(a.report_period)
  );
  const target = period
    ? sorted.find(p => p.report_period === period)
    : sorted[0];
  if (!target) return [];

  const reportPeriod = target.report_period;
  const seenKeys     = new Set<string>();
  const results: ParsedSegment[] = [];

  for (const item of target.items) {
    // ── 1. Filter by metric ──────────────────────────────────────────────────
    if (item.name !== metric_name) continue;

    const segs = item.segments;

    // ── 2. Filter by axis + enforce leaf constraint ──────────────────────────
    if (leaf_only) {
      // Items with no segments are dimension-less totals → aggregate
      if (segs.length === 0) continue;
      // Items spanning multiple segment tags → cross-dimensional → aggregate
      if (segs.length !== 1) continue;
      if (segs[0].axis !== axis) continue;
    } else {
      const axisSegs = segs.filter(s => s.axis === axis);
      if (axisSegs.length !== 1) continue;
    }

    const seg = segs[0];
    if (!seg.key) continue;

    // ── 3. Deduplication by key (NEVER by label) ────────────────────────────
    if (dedupe && seenKeys.has(seg.key)) continue;
    seenKeys.add(seg.key);

    results.push({
      key:    seg.key,
      label:  seg.label || seg.key,
      value:  item.amount,
      axis,
      metric: metric_name,
      period: reportPeriod,
    });
  }

  // ── 4. Remove aggregate totals ───────────────────────────────────────────
  return leaf_only ? removeAggregates(results) : results;
}


function removeAggregates(segs: ParsedSegment[]): ParsedSegment[] {
  if (segs.length < 2) return segs;
  const total = segs.reduce((s, x) => s + x.value, 0);
  return segs.filter(seg => {
    const othersSum = total - seg.value;
    // If this value ≈ sum of all others, it's a grand-total row → drop it
    return !(othersSum > 0 && Math.abs(seg.value - othersSum) / othersSum < AGG_TOLERANCE);
  });
}


// ── Discovery helpers ─────────────────────────────────────────────────────────

/** Return all unique axis identifiers present in the latest period. */
export function discoverAxes(periods: SegmentedRevenuePeriod[]): string[] {
  if (!periods.length) return [];
  const latest = latestPeriod(periods);
  const axes   = new Set<string>();
  for (const item of latest.items) {
    for (const seg of item.segments) {
      if (seg.axis) axes.add(seg.axis);
    }
  }
  return [...axes];
}

/**
 * Return the metric name (item.name) with the most leaf items for *axis*
 * in the latest period.  Used to auto-select the primary metric.
 */
export function dominantMetric(
  periods: SegmentedRevenuePeriod[],
  axis:    string,
): string | null {
  if (!periods.length) return null;
  const latest = latestPeriod(periods);
  const counts = new Map<string, number>();
  for (const item of latest.items) {
    if (item.segments.length === 1 && item.segments[0].axis === axis) {
      counts.set(item.name, (counts.get(item.name) ?? 0) + 1);
    }
  }
  if (!counts.size) return null;
  return [...counts.entries()].sort((a, b) => b[1] - a[1])[0][0];
}

function latestPeriod(periods: SegmentedRevenuePeriod[]): SegmentedRevenuePeriod {
  return [...periods].sort((a, b) =>
    b.report_period.localeCompare(a.report_period)
  )[0];
}


// ── Axis label map (generic fallback for unknown axes) ────────────────────────

export interface AxisLabels {
  title:     string;   // section heading
  donut:     string;   // donut chart sub-title
  hist:      string;   // historical bar chart sub-title
}

const AXIS_LABEL_MAP: Record<string, AxisLabels> = {
  "srt:ProductOrServiceAxis": {
    title: "Product & Service Breakdown",
    donut: "Revenue by Product",
    hist:  "Historical Revenue by Product",
  },
  "srt:StatementGeographicalAxis": {
    title: "Geographic Breakdown",
    donut: "Revenue by Geography",
    hist:  "Historical Revenue by Geography",
  },
  "us-gaap:StatementBusinessSegmentsAxis": {
    title: "Business Segment Breakdown",
    donut: "Revenue by Segment",
    hist:  "Historical Revenue by Segment",
  },
};

/**
 * Return human-readable labels for a given axis URI.
 * Falls back gracefully for unknown axes using the local part of the CURIE.
 */
export function axisLabels(axis: string): AxisLabels {
  if (axis in AXIS_LABEL_MAP) return AXIS_LABEL_MAP[axis];
  // Derive a reasonable label from the XBRL CURIE  e.g. "foo:BarBazAxis" → "Bar Baz"
  const local     = axis.split(":").pop() ?? axis;
  const humanized = local.replace(/Axis$/, "").replace(/([A-Z])/g, " $1").trim();
  return {
    title: `${humanized} Breakdown`,
    donut: `Revenue by ${humanized}`,
    hist:  `Historical Revenue by ${humanized}`,
  };
}
